fix(ivy): don't match directives against attribute bindings

crisbeto opened the pull request:
Fixes Ivy matching directives against attribute bindings (e.g. `[attr.some-directive]="foo"`). Works by excluding attribute bindings from the attributes array during compilation. This has the added benefit of generating less code.

**Note:** My initial approach to implementing this was to have a different marker for attribute bindings so that they can be ignored when matching directives, however as I was implementing it I realized that the attributes in that array were only used for directive matching (as far as I could tell). I decided to drop the attribute bindings completely, because it results in less generated code.

pkozlowski-opensource commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -199,7 +199,7 @@ export const enum AttributeMarker {
    * var _c1 = ['moo', 'car', AttributeMarker.I18n, 'foo', 'bar'];
    */
-  I18n,
+  I18n = 6,
```
👍

pkozlowski-opensource commented on packages/core/test/acceptance/directive_spec.ts:
```diff
@@ -202,6 +202,29 @@ describe('directives', () => {
+        template: `<p [attr.dir]="getDir()"></p>`,
+      })
+      class MyComp {
+        getDir() { return 'auto'; }
```
Do we need a getter here? We could just use a constant in a binding (see below).

pkozlowski-opensource commented on packages/core/test/acceptance/directive_spec.ts:
```diff
@@ -202,6 +202,29 @@ describe('directives', () => {
+      const fixture = TestBed.createComponent(MyComp);
+      fixture.detectChanges();
+
+      expect(calls).toEqual([]);
```
super nit: I would write this test slightly differently just to make sure that a directive matches sth. So I would probably do a template like `<p [attr.dir]="'auto'"></p><p dir="auto"></dir>` and check that there is one call (or just try to get a directive instance from an injector)

crisbeto commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -333,7 +333,7 @@ describe('i18n support in the view compiler', () => {
 
       const output = `
-        const $_c0$ = [${AttributeMarker.Bindings}, "title", "label"];
+        const $_c0$ = [3, "title"];
```
Note: this change really shouldn't be necessary, but it was tripping up the linter which was asking me to insert some invalid code further down. See:

![Continuous_Integration_and_Deployment_-_Google_Chr_2019-07-13_12-28-24](https://user-images.githubusercontent.com/4450522/61170839-33c4d000-a56f-11e9-9a83-3790dbe76208.png)

pkozlowski-opensource reviewed (approved):
LGTM! Thnx for fixing, I can't believe that we had it so broken for so long :-)

crisbeto committed:
fix(ivy): don't match directives against attribute bindings

Fixes Ivy matching directives against attribute bindings (e.g. `[attr.some-directive]="foo"`). Works by excluding attribute bindings from the attributes array during compilation. This has the added benefit of generating less code.

**Note:** My initial approach to implementing this was to have a different marker for attribute bindings so that they can be ignored when matching directives, however as I was implementing it I realized that the attributes in that array were only used for directive matching (as far as I could tell). I decided to drop the attribute bindings completely, because it results in less generated code.

crisbeto commented:
The Codefresh failure looks like a flake.

AndrewKushnir reviewed (approved):
LGTM, thanks for the fix @crisbeto!

AndrewKushnir commented:
- [VE Presubmit](https://test.corp.google.com/ui#id=OCL:257966242:BASE:257966240:1563040584778:432f6d9)
- [Ivy Presubmit](https://test.corp.google.com/ui#id=OCL:257966242:BASE:257966276:1563040661196:4b9e8c93)

AndrewKushnir commented:
FYI, VE presubmit is successful, Ivy presubmit looks normal 👍

kara reviewed (approved):
LGTM

kara commented:
merge-assistance: global approval
